// @innovapptive.com/dynamictable/esm2020/lib/avatar.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class AvatarComponent {
    ngOnInit() { }
    getText() {
        if (!this.text) {
            return;
        }
        let matches = this.text.charAt(0);
        return matches.toUpperCase();
    }
}
AvatarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AvatarComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
AvatarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.3", type: AvatarComponent, selector: "avatar", inputs: { text: "text" }, ngImport: i0, template: ` <span class="avatar">{{ getText() }}</span> `, isInline: true, styles: [".avatar{width:40px;height:40px;background:#cacaca;border-radius:5px;display:inline-grid;text-align:center;align-items:center;color:#88888a;font-size:150%;font-weight:500}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: AvatarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'avatar',
                    template: ` <span class="avatar">{{ getText() }}</span> `,
                    styles: [
                        `
      .avatar {
        width: 40px;
        height: 40px;
        background: #cacaca;
        border-radius: 5px;
        display: inline-grid;
        text-align: center;
        align-items: center;
        color: #88888a;
        font-size: 150%;
        font-weight: 500;
      }
    `
                    ]
                }]
        }], propDecorators: { text: [{
                type: Input
            }] } });
