// @innovapptive.com/dynamictable/esm2020/lib/dynamictable.service.mjs
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DynamictableFilterService {
    constructor() {
        this.tableLeftPanelEventEmitter = new EventEmitter();
    }
    emitEvent(data) {
        this.tableLeftPanelEventEmitter.emit(data);
    }
    getEventEmitter() {
        return this.tableLeftPanelEventEmitter;
    }
}
DynamictableFilterService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableFilterService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DynamictableFilterService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableFilterService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableFilterService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
export class DynamictableScrollerService {
    constructor() {
        this.scrollEventEmitter = new EventEmitter();
    }
    emitEvent(data) {
        this.scrollEventEmitter.emit(data);
    }
    getEventEmitter() {
        return this.scrollEventEmitter;
    }
}
DynamictableScrollerService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableScrollerService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DynamictableScrollerService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableScrollerService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableScrollerService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });
