// @innovapptive.com/dynamictable/esm2020/lib/dynamictable.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DynamictableComponent } from './dynamictable.component';
import { SharedMaterialModule } from './shared-material-module';
import { FormsModule } from '@angular/forms';
import { TableFilterPanelComponent } from './table-filter-panel/table-filter-panel.component';
import { DynamictableFilterService } from './dynamictable.service';
import { DynamicTypePipe } from './dynamic-type-pipe/dynamic-type-pipe';
import { AvatarComponent } from './avatar.component';
import { TimeAgoPipe } from './time-ago-type-pipe/time-ago.pipe';
import * as i0 from "@angular/core";
export class DynamictableModule {
}
DynamictableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DynamictableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableModule, declarations: [DynamictableComponent,
        TableFilterPanelComponent,
        DynamicTypePipe,
        AvatarComponent,
        TimeAgoPipe], imports: [CommonModule, FormsModule, SharedMaterialModule], exports: [DynamictableComponent, TableFilterPanelComponent] });
DynamictableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableModule, providers: [DynamictableFilterService, TimeAgoPipe], imports: [[CommonModule, FormsModule, SharedMaterialModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamictableModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DynamictableComponent,
                        TableFilterPanelComponent,
                        DynamicTypePipe,
                        AvatarComponent,
                        TimeAgoPipe
                    ],
                    imports: [CommonModule, FormsModule, SharedMaterialModule],
                    exports: [DynamictableComponent, TableFilterPanelComponent],
                    providers: [DynamictableFilterService, TimeAgoPipe]
                }]
        }] });
