// @innovapptive.com/dynamictable/esm2020/lib/dynamic-type-pipe/dynamic-type-pipe.mjs
import { Pipe } from '@angular/core';
import { DatePipe } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "../time-ago-type-pipe/time-ago.pipe";
export class DynamicTypePipe {
    constructor(timeAgoPipe) {
        this.timeAgoPipe = timeAgoPipe;
    }
    transform(val, pipeType) {
        switch (pipeType) {
            case 'date':
                val = new DatePipe('en-US').transform(val, 'short');
                break;
            case 'timeAgo':
                val = this.timeAgoPipe.transform(val);
                break;
            default:
            // do nothing
        }
        return val;
    }
}
DynamicTypePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamicTypePipe, deps: [{ token: i1.TimeAgoPipe }], target: i0.ɵɵFactoryTarget.Pipe });
DynamicTypePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamicTypePipe, name: "dynamicTypePipe" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.3", ngImport: i0, type: DynamicTypePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'dynamicTypePipe'
                }]
        }], ctorParameters: function () { return [{ type: i1.TimeAgoPipe }]; } });
